### PR TITLE
Switch from AbstractDifferentiation to DifferentiationInterface

### PR DESCRIPTION
**Pull request by gdalle**
Following our discussion per email, this PR proposes a switch from AbstractDifferentiation.jl to DifferentiationInterface.jl, which is becoming the new standard in the ecosystem.

- [x] Modify `Project.toml` files and imports
- [x] Replace `SomethingBackend()` with `AutoSomething()`
- [x] Replace `value_and_gradient_closure` with `value_and_gradient` (unclear how performance is affected)
- [x] Update documentation and README
- [ ] Add [preparation mechanism](https://gdalle.github.io/DifferentiationInterface.jl/DifferentiationInterface/stable/operators/#Preparation): available on another branch but not sure we want it because if the function contains value-dependent control flow, preparation is not appropriate

### CODE REVIEW

**Comment by gdalle** [comment]
Any idea why the benchmarks fail? Or opinions on the preparation mechanism, which I could add to the present PR?

**Comment by lostella** [comment]
@gdalle thanks for this! I need to take a deeper look into why the benchmarks fail, seems like something ends up in an endless loop, so maybe it's something to fix upstream.

Regarding preparation: maybe it can be added separately, with some mechanism to make it optional

**Comment by lostella** [comment]
@gdalle #94 fixed the benchmark, apologies for the conflicts

**Comment by gdalle** [comment]
Fixed the conflicts, this should be good to go! We can always discuss preparation in a separate PR

**Comment by gdalle** [comment]
@lostella can you authorize the workflows?

**Comment by gdalle** [comment]
It seems the benchmarks worked for the PR branch but failed for the main branch?

**Comment by lostella** [comment]
Yes I think this is expected, since the change is breaking and the benchmark code won't work with the previous commit, right? I don't know of a better method to organize benchmarks for these cases.

I think it's fine, nothing to worry about. I'll just take a manual look at performance before merging, otherwise this looks good, thank you!

File: Project.toml [hunk 1]
```diff
@@ -3,13 +3,15 @@ uuid = "140ffc9f-1907-541a-a177-7475e0a401e9"
 version = "0.6.0"
```
Comment (lostella) [inline Project.toml:3]:
```suggestion
version = "0.7.0"
```

**Comment by gdalle** [comment]
Hi @lostella, just popping by to see if you need anything else from me

**Review by lostella — APPROVED** [review]

**Comment by gdalle** [comment]
While we're at it, do you want to try to update to DI v0.6, which just landed?

**Comment by lostella** [comment]
> While we're at it, do you want to try to update to DI v0.6, which just landed?

Makes sense!

**Pull request merged**